From c11b48dea976a3f6237dbcd6e133c872094ed026 Mon Sep 17 00:00:00 2001
From: Basheer Ahmed Muddebihal <basheer.ahmed.muddebihal@linux.intel.com>
Date: Tue, 19 Jul 2022 13:53:03 -0700
Subject: [PATCH] fpga: dfl: Add GUID support in mod device table

In the modalias table, dfl devices use the type id and feature id to
create the dfl device string entry "dfl:t0000f0009*". This patch adds
support for the use of a GUID for dfl devices. Device drivers may match
on type id and feature id, on the GUID, or on a combination of the type id,
feature id and GUID.

If the feature id and GUID are both non-zero, then the modalias string is:
"dfl:t0000f0009g{525074DC-8985-46E2-8057-A307DC18A509}*"
If the feature id is zero and the GUID is non-zero, then the string is:
"dfl:t*f*g{525074DC-8985-46E2-8057-A307DC18A509}*"
If the feature id is non-zero and the GUID is zero, then the string is
"dfl:t0000f0009*"

Signed-off-by: Basheer Ahmed Muddebihal <basheer.ahmed.muddebihal@linux.intel.com>
---
 include/linux/mod_devicetable.h   |  1 +
 scripts/mod/devicetable-offsets.c |  1 +
 scripts/mod/file2alias.c          | 22 ++++++++++++++++++++--
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/include/linux/mod_devicetable.h b/include/linux/mod_devicetable.h
index b0678b093cb2..018f5fb583ec 100644
--- a/include/linux/mod_devicetable.h
+++ b/include/linux/mod_devicetable.h
@@ -908,6 +908,7 @@ struct ssam_device_id {
 struct dfl_device_id {
 	__u16 type;
 	__u16 feature_id;
+	guid_t guid;
 	kernel_ulong_t driver_data;
 };
 
diff --git a/scripts/mod/devicetable-offsets.c b/scripts/mod/devicetable-offsets.c
index abe65f8968dd..460a76f5b2c0 100644
--- a/scripts/mod/devicetable-offsets.c
+++ b/scripts/mod/devicetable-offsets.c
@@ -258,6 +258,7 @@ int main(void)
 	DEVID(dfl_device_id);
 	DEVID_FIELD(dfl_device_id, type);
 	DEVID_FIELD(dfl_device_id, feature_id);
+	DEVID_FIELD(dfl_device_id, guid);
 
 	DEVID(ishtp_device_id);
 	DEVID_FIELD(ishtp_device_id, guid);
diff --git a/scripts/mod/file2alias.c b/scripts/mod/file2alias.c
index 7056751c29b1..981b7707d7f6 100644
--- a/scripts/mod/file2alias.c
+++ b/scripts/mod/file2alias.c
@@ -1440,14 +1440,32 @@ static int do_ssam_entry(const char *filename, void *symval, char *alias)
 	return 1;
 }
 
-/* Looks like: dfl:tNfN */
+/* Looks like: dfl:tNfNg{guid} */
 static int do_dfl_entry(const char *filename, void *symval, char *alias)
 {
+	int guid_cmp_val;
+	guid_t null_guid = {0};
 	DEF_FIELD(symval, dfl_device_id, type);
 	DEF_FIELD(symval, dfl_device_id, feature_id);
+	DEF_FIELD(symval, dfl_device_id, guid);
 
-	sprintf(alias, "dfl:t%04Xf%04X", type, feature_id);
+	guid_cmp_val = memcmp(&null_guid, &guid, sizeof(guid_t));
 
+	if (feature_id == 0 && guid_cmp_val == 0) {
+		warn("Invalid dfl Device ID for in '%s'\n", filename);
+		return 0;
+	}
+
+	if (feature_id == 0)
+		strcpy(alias, "dfl:t*f*");
+	else
+		snprintf(alias, ALIAS_SIZE, "dfl:t%04Xf%04X", type, feature_id);
+
+	if (guid_cmp_val) {
+		strcat(alias + strlen(alias), "g{");
+		add_guid(alias, guid);
+		strcat(alias + strlen(alias), "}");
+	}
 	add_wildcard(alias);
 	return 1;
 }
